Feature/animation rendering

petersteneteg opened the pull request:
Changes proposed in this PR:
 * Make the rendering mode in the animation editor usable

petersteneteg committed:
Animation: better handling of rendering animations

petersteneteg committed:
Core: added a processEvent callback in inviwoapplication for eventloops to hook into used by the Animation module

martinfalk commented on modules/animation/src/datastructures/invalidationtrack.cpp:
```diff
@@ -0,0 +1,93 @@
+}
+std::string InvalidationTrack::getClassIdentifier() const { return classIdentifier(); }
+
+AnimationTimeState InvalidationTrack::operator()(Seconds from, Seconds to,
```
`from` unused?

petersteneteg committed:
Animation: review fix

Inviwo-Team committed:
Jenkins: Format fixes

ResearchDaniel commented on modules/animation/include/modules/animation/datastructures/invalidationtrack.h:
```diff
@@ -0,0 +1,103 @@
+    std::string processorId;
+};
+
+class IVW_MODULE_ANIMATION_API InvalidationTrack : public BaseTrack<InvalidationKeyframeSequence> {
```
I think that the connection to what is invalidated could be more explicit as both properties and processors can be invalidated.
I am also not sure about the use case for this track. If a processor has not been invalidated there will not be any change in its output. Care to elaborate on this by documenting the class?
```suggestion
class IVW_MODULE_ANIMATION_API ProcessorInvalidationTrack : public BaseTrack<InvalidationKeyframeSequence> {
```

ResearchDaniel replied:
If the name is changed I also think that corresponding changes should be made to the KeyFrame and the KeyFrameSequence

ResearchDaniel commented on modules/animation/src/animationcontroller.cpp:
```diff
@@ -135,105 +112,79 @@ AnimationController::AnimationController(Animation& animation, InviwoApplication
+                 }
+                 return state;
+             }())
+    , renderFPS("renderFPS", "Frames per Second", 100, 2, 1000000, 1,
```
Ranges need to be adapted because the meaning of this property changed
```suggestion
    , renderFPS("renderFPS", "Frames per Second", 24.0,
                      {1, ConstraintBehavior::Immutable},
                      {60, ConstraintBehavior::Editable}, 1.0
```

ResearchDaniel commented on modules/animation/src/animationcontroller.cpp:
```diff
@@ -87,42 +88,18 @@ class Layer;
-          [this](const double& value) { deltaTime_ = Seconds(1.0 / value); },
-          {0.001, ConstraintBehavior::Immutable}, {1000.0, ConstraintBehavior::Immutable}, 1.0,
-          InvalidationLevel::InvalidOutput, PropertySemantics::Text)
+    , framesPerSecond("PlayFramesPerSecond", "Frames per Second", 25.0,
```
Out of scope of PR, but not sure why it was 25 before. I think that 24, 30, or 60 would make most sense.
```suggestion
    , framesPerSecond("PlayFramesPerSecond", "Frames per Second", 24.0,
```

ResearchDaniel commented on modules/animation/src/animationcontroller.cpp:
```diff
@@ -469,46 +303,88 @@ void AnimationController::tick() {
+        if (p->isSink()) {
+            if (auto exporter = dynamic_cast<Exporter*>(p)) {
+                auto base = renderBaseName.get();
+                replaceInString(base, "UPN", p->getIdentifier());
```
Where does this UPN come from and what is its purpose?

petersteneteg replied:
Think it's an old invention by Rickard. it means "use processor name"
There are probably better ways to do that... but that's how we have done it for a long time...

ResearchDaniel commented on modules/animation/src/animationcontroller.cpp:
```diff
@@ -469,46 +303,88 @@ void AnimationController::tick() {
+
+        lock.reset();
+
+        while (app_->getProcessorNetwork()->runningBackgroundJobs() > 0) {
```
This functionality seems to be a reocurring pattern.
See for example `waitForNetwork` in the python API.
Should we expose this in the application or is it already there through `waitForPool` (should processEvents be included in waitForPool?)? 
If not, what is the difference between these ways of waiting?

ResearchDaniel commented on modules/animation/src/animationcontroller.cpp:
```diff
@@ -469,46 +303,88 @@ void AnimationController::tick() {
+                       std::chrono::high_resolution_clock::now() - start)
+                       .count();
+    LogInfo(fmt::format("Rendered {} frames in {:.3f} seconds, {:.3f} per frame", numFrames,
+                        seconds, seconds / numFrames));
```
Does use of int not produce a warning?
```suggestion
                        seconds, seconds / static_cast<double>(numFrames)));
```

petersteneteg replied:
I think double / int should be fine

petersteneteg committed:
Animation: review fixes

# Conflicts:
#	modules/animation/include/modules/animation/datastructures/invalidationtrack.h

Inviwo-Team committed:
Jenkins: Format fixes

ResearchDaniel commented on modules/animation/src/animationcontroller.cpp:
```diff
@@ -91,67 +91,65 @@ namespace animation {
-                     {{"FullTimeWindow", "Play full animation", 0},
+                     {
+                         {"FullTimeWindow", "Play full animation",  0},
+                         {"UserTimeWindow", "Selected time window", 1}
```
Not sure how much is given by the context, but it might be better with:
“Play selected”
Or 
“Play selected frames”

ResearchDaniel commented on modules/animation/src/animationcontroller.cpp:
```diff
@@ -425,6 +423,6 @@ Seconds AnimationController::deltaTime() const {
 
 Seconds AnimationController::getCurrentTime() const { return currentTime_; }
 
-}  // namespace animation
```
Is this a new format setting?
Personally, I think that there should be a space

petersteneteg replied:
yes, I have added it back. had some local version update of clang-format which did the wrong thing...

ResearchDaniel commented on modules/animationqt/src/sequenceeditor/invalidationsequenceeditor.cpp:
```diff
@@ -130,15 +130,15 @@ InvalidationSequenceEditor::InvalidationSequenceEditor(KeyframeSequence& sequenc
-    subLayout->addWidget(new QLabel("Processor Identifier"), 0, 0);
-    subLayout->addWidget(processor_, 0, 1);
+    path_ = new QLineEdit();
+    path_->setToolTip("Invalidate the specified processor / property at each network evaluation");
```
Suggest to add an example as people might not know what a path is/looks like:
(processorId.propertyId, e.g., Canvas or Canvas.saveLayer)

ResearchDaniel reviewed (dismissed):
Space after comment “//“ should not be removed?

petersteneteg committed:
Animation: review fixes
